File: Semestr III/Scala, Ocaml, Java/Paradygmaty programowania/Ocaml/lista5.ml
(*Zad 1*)

type 'a llist = LNil | LCons of 'a * (unit -> 'a llist);;

let lrepeat llista k =
	let rec helper (result,iter) =
		match (result, iter) with
		| (LNil,_) -> LNil
		| (LCons(_,t), 0) -> helper(t(),k)
		|	(LCons(h,_), _) -> LCons(h, fun() -> helper(result,iter - 1))
	in helper(llista, k)
	;;

let lLista1 = (LCons(1, function () -> LCons(2, function () ->LCons(3, function () -> LCons(4, function () ->LCons(5, function () -> LNil))))));;
let x = lrepeat lLista1 3;;
ltake(15,x)
int list = [1; 1; 1; 2; 2; 2; 3; 3; 3; 4; 4; 4; 5; 5; 5]

(*lfrom*)
let rec lfrom k =
    LCons (k, function () -> lfrom (k+1))
    ;;

(*ltake*)
 let rec ltake = function
    | (0, _) -> [] 
    | (_, LNil) -> []
    | (n, LCons(x,xf)) -> x::ltake(n-1, xf())
    ;;

(*toLazyList*)
let rec toLazyList = function
    | [] -> LNil
    | h::t -> LCons(h, function () -> toLazyList t)
    ;;



(*Zad 2*)

let lfib =
	let rec fib a b = 
		LCons(a, fun() -> fib b (a + b))
	in fib 1 1
	;;



(*Zad 3*)

type 'a lBT = LEmpty | LNode of  'a * (unit ->'a lBT) * (unit -> 'a lBT);;



(*a)*)

let rec lTree n =
	LNode(n, (fun () -> lTree(2*n)), (fun () -> lTree(2*n+1)))
	;;



(*b)*)

let breadthlBT lBT =
	let rec recTree = function
		| [] -> LNil
		| LEmpty :: tl         -> recTree tl
		| LNode(v, l, r) :: tl -> LCons(v, fun () -> recTree (tl @ [l(); r()]))
	in recTree [lBT]

File: Semestr III/Scala, Ocaml, Java/Paradygmaty programowania/Scala/lista5.scala
//Zad 1

def lrepeat[A](llista: Stream[A]) (k: Int): Stream[A] = {
  def helper[A](result: Stream[A], iter: Int): Stream[A] =
    (result, iter) match {
    case (Stream.Empty, _) => Stream.Empty
    case ((_#::t), 0) => helper(t, k)
    case ((h#::_), _) => h #:: helper(result, iter - 1)
  }
  helper (llista, k)
}
val s1 = Stream(1,2,3,4,5)

val x = lrepeat (s1) (3)

x.toList
res2: List[Int] = List(1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5)



//Zad 2

val lfib = {
  def fib(a: Int, b: Int): Stream[Int] =
    a #:: fib(b, b+a)
    fib(1,1)
}

lfib.toList


//Zad 3
sealed trait lBT[+A]
case object LEmpty  extends lBT[Nothing]
case class LNode[+A](elem:A, left:()=>lBT[A], right:()=>lBT[A]) extends lBT[A]

//a)

def lTree(n: Int): lBT[Int] =
  LNode(n, () => lTree(2*n), () => lTree(2*n+1))
  
  
  
//b)
  
def breadthlBT[A] (lBtree: lBT[A]) = {
  def recTree (kolejka: List[lBT[A]]): Stream[A] = 
    kolejka match {
      case Nil => Stream.Empty
      case (LEmpty::tl) => recTree(tl)
      case (LNode(v,l,r) :: tl) => v #:: recTree(tl ++ List(l(), r()))
  }
  recTree(List(lBtree))
}
  